### PR TITLE
Measurement: Fix macos crashes caused by unhandled exceptions reaching qt

### PR DESCRIPTION
**Pull request by hyarion**
As described in https://github.com/FreeCAD/FreeCAD/issues/16905, macos crashes when exceptions are reaching qt.

The crash this PR fixes is described this comment: https://github.com/FreeCAD/FreeCAD/issues/16888#issuecomment-2381031659

The reason why FreeCAD crashes is:
1. OCCT throws an exception [here](https://github.com/FreeCAD/FreeCAD/blob/6e2cd4e73303f817e4bf5e908abe39030b7c5fb3/src/Mod/Measure/App/Measurement.cpp#L361)
2. The exception is not caught in `QuickMeasure::processSelection()`

This PR adds a catch for OCCT exceptions as well as a generic for unknown exceptions.

Note, this doesn't fix the bug, but it fixes the crash.

### CODE REVIEW

**Comment by PaddleStroke** [comment]
So I was looking into this before seeing this PR.
In Measurement.cpp there're also 2 places where there's no `catch(...)` block. My understanding on this topic is a bit limited, so can you please advise if they should also be added?

**Comment by hyarion** [comment]
@PaddleStroke 
We must be really careful and catch all exceptions so reaches qt code.
In linux exceptions passes through down to GuiApplication::notify without a problem, but on macOS (arm64) it crashes.

More information an links to a Qt bug on the topic can be found [here](https://github.com/FreeCAD/FreeCAD/issues/16905)

What this means for us is that all places where a linux user might get a error message `Unhandled {some exception}  caught in GUIApplication::notify.` must be handled earlier. AIRCAP wrote a good comment about this [here](https://github.com/FreeCAD/FreeCAD/issues/16905#issuecomment-2381337120).

<details><summary>For this bug, the callstack looked follows 
 (click to expand)</summary>
<p>

---

At the top of the stack is where the thread was killed due to unhandled exception (it never goes through qt on macos:arm64, on linux it does and is handled)

* `libsystem_kernel.dylib`__pthread_kill + 8`
* `libsystem_pthread.dylib`pthread_kill + 288`
* `libsystem_c.dylib`abort + 180`
* `libc++abi.dylib`abort_message + 132`
* `libc++abi.dylib`demangling_terminate_handler() + 348`
* `libobjc.A.dylib`_objc_terminate() + 160`
* `libc++abi.dylib`std::__terminate(void (*)()) + 16`
* `libc++abi.dylib`__cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 88`
* `libc++abi.dylib`__cxa_throw + 308`

The exception was thrown here:
* `Measure.so`TopoDS::Edge(S=0x000000016fdfb770) at TopoDS.lxx:71:3`
* `Measure.so`Measure::Measurement::length(this=0x000000012cdbff40) const at Measurement.cpp:361:43`
* `MeasureGui.so`MeasureGui::QuickMeasure::printResult(this=0x000000012cdbf840) at QuickMeasure.cpp:208:42`
* `MeasureGui.so`MeasureGui::QuickMeasure::tryMeasureSelection(this=0x000000012cdbf840) at QuickMeasure.cpp:101:5`
* `MeasureGui.so`MeasureGui::QuickMeasure::processSelection(this=0x000000012cdbf840) at QuickMeasure.cpp:81:13`
* `MeasureGui.so`QtPrivate::FunctorCall<QtPrivate::IndexesList<>, QtPrivate::List<>, void, void (MeasureGui::QuickMeasure::*)()>::call(f=(MeasureGui.so`MeasureGui::QuickMeasure::processSelection() at QuickMeasure.cpp:77), o=0x000000012cdbf840, arg=0x000000016fdfbe48) at qobjectdefs_impl.h:152:13`
* `MeasureGui.so`void QtPrivate::FunctionPointer<void (MeasureGui::QuickMeasure::*)()>::call<QtPrivate::List<>, void>(f=(MeasureGui.so`MeasureGui::QuickMeasure::processSelection() at QuickMeasure.cpp:77), o=0x000000012cdbf840, arg=0x000000016fdfbe48) at qobjectdefs_impl.h:185:13`

MeasureGui was invoked using the slot here:
* `MeasureGui.so`QtPrivate::QSlotObject<void (MeasureGui::QuickMeasure::*)(), QtPrivate::List<>, void>::impl(which=1, this_=0x000000012cdc0f00, r=0x000000012cdbf840, a=0x000000016fdfbe48, ret=0x0000000000000000) at qobjectdefs_impl.h:418:17`
* `libQt5Core.5.15.8.dylib`___lldb_unnamed_symbol16968 + 820`

The slot was called by a timedout QTimer:
* `libQt5Core.5.15.8.dylib`QTimer::timeout(QTimer::QPrivateSignal) + 52`
* `libQt5Core.5.15.8.dylib`QObject::event(QEvent*) + 104`

The event was sent by QApplication's notify:
* `libQt5Widgets.5.15.8.dylib`QApplicationPrivate::notify_helper(QObject*, QEvent*) + 232`
* `libQt5Widgets.5.15.8.dylib`QApplication::notify(QObject*, QEvent*) + 472`

Here is Gui::GuiApplication::notify where we are trying to catch the event, but it doesn't work as explained in the Qt ticket
* `libFreeCADGui.dylib`Gui::GUIApplication::notify(this=0x000000016fdfe2b8, receiver=0x000000012cdc0e60, event=0x000000016fdfc430) at GuiApplication.cpp:83:34`
* `libQt5Core.5.15.8.dylib`QCoreApplication::notifyInternal2(QObject*, QEvent*) + 176`

The tail here isn't really that interesting, but I'll keep it for the record:
* `libQt5Core.5.15.8.dylib`QTimerInfoList::activateTimers() + 520`
* `libqcocoa.dylib`___lldb_unnamed_symbol3667 + 20`
* `libqcocoa.dylib`___lldb_unnamed_symbol3666 + 16`
* `CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 28`
* `CoreFoundation`__CFRunLoopDoSource0 + 176`
* `CoreFoundation`__CFRunLoopDoSources0 + 244`
* `CoreFoundation`__CFRunLoopRun + 828`
* `CoreFoundation`CFRunLoopRunSpecific + 608`
* `HIToolbox`RunCurrentEventLoopInMode + 292`
* `HIToolbox`ReceiveNextEventCommon + 648`
* `HIToolbox`_BlockUntilNextEventMatchingListInModeWithFilter + 76`
* `AppKit`_DPSNextEvent + 660`
* `AppKit`-[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 700`
* `AppKit`-[NSApplication run] + 476`
* `libqcocoa.dylib`___lldb_unnamed_symbol3677 + 968`
* `libQt5Core.5.15.8.dylib`QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 312`
* `libQt5Core.5.15.8.dylib`QCoreApplication::exec() + 120`
* `libFreeCADGui.dylib`(anonymous namespace)::tryRunEventLoop(mainApp=0x000000016fdfe2b8) at Application.cpp:2207:5`
* `libFreeCADGui.dylib`(anonymous namespace)::runEventLoop(mainApp=0x000000016fdfe2b8) at Application.cpp:2227:9`
* `libFreeCADGui.dylib`Gui::Application::runApplication() at Application.cpp:2304:5`
* `FreeCAD`main(argc=3, argv=0x000000016fdfeef8) at MainGui.cpp:295:13`
* `dyld`start + 2360`

---

</p>
</details> 

As can be seen in the call stack, the point of entry was processSelection and since there already was a try-catch-structure there, I simply added:
* a case for the base OCCT exception (Standard_Failore), and
* a case for any other type of exception

In other words, I moved up the error handling from GuiApplication::notify to the last point before the exception would reach Qt which would cause a crash.

I haven't checked if there are more slots which could be called. If so those needs to be handled as well.


To answer your question,
I think it would be best to add exception handling in _all_ slots, but that discussion would be better to have in [this issue](https://github.com/FreeCAD/FreeCAD/issues/16905).

I looked at the two try-catch structures in measurement and it seems like they are utility functions, not slots, so it might not be needed.

**Comment by chennes** [comment]
Backported to 1.0
